File: expandable-raid.php
#!/usr/bin/php
<?

/* vim set noexpandtab */

/**
 * 
 * expandable-raid.php
 * 
 * Use Linux RAID and LVM together to add disks to your RAID array without stopping it
 * 
 * https://github.com/stevecoug/expandable-raid
 * 
 * Copyright (c) 2013 Steve Meyers
 * 
 * This Source Code Form is subject to the terms of the Mozilla Public License,
 * v. 2.0. If a copy of the MPL was not distributed with this file, You can
 * obtain one at http://mozilla.org/MPL/2.0/.
 *
 */


$mode = false;
$partitions = [];
$volgroup = false;
$raiddev = false;
$chunk = 64;
$level = 5;
$layout = false;
$dryrun = false;

try {
	for ($i = 1; $i < $argc; $i++) {
		switch ($argv[$i]) {
			case "-c":
			case "--create":
				$mode = "create";
			break;
			
			case "-x":
			case "--extend":
				$mode = "extend";
			break;
			
			case "--remove":
				$mode = "remove";
			break;
			
			case "-v":
			case "--vg":
			case "--volgroup":
				$volgroup = $argv[++$i];
			break;
			
			case "-l":
			case "--level":
				$level = intval($argv[++$i]);
				if (!in_array($level, [ 0, 1, 5, 10 ])) throw new Exception("Invalid RAID level: $level");
			break;
			
			case "-r":
			case "--raid":
				$raiddev = $argv[++$i];
				if ($raiddev[0] !== "/") $raiddev = "/dev/$raiddev";
				if (!preg_match('|^/dev/md[0-9]+$|', $raiddev)) throw new Exception("Invalid RAID device: $raiddev");
			break;
			
			case "-p":
			case "--partition":
			case "--partitions":
				$partitions = [];
				$tmp = explode(",", $argv[++$i]);
				foreach ($tmp as $part) {
					if ($part[0] !== "/") $part = "/dev/$part";
					$partitions[] = $part;
				}
			break;
			
			case "--chunk":
				$chunk = intval($argv[++$i]);
				if ($chunk < 1 || $chunk > 1024) throw new Exception("Chunk size must be between 1-1024");
			break;
			
			case "--layout":
				$layout = $argv[++$i];
			break;
			
			case "--dry-run":
				$dryrun = true;
			break;
			
			default:
				throw new Exception("Invalid argument: ".$argv[$i]);
			break;
		}
	}
	
	if ($mode === false) throw new Exception("You must specify either --create or --extend");
	
	if ($volgroup === false) throw new Exception("Volume group must be specified");
	if (count($partitions) == 0 && in_array($mode, [ "create", "extend" ])) throw new Exception("Partitions must be specified");
	
	if ($mode == "extend") {
		if ($raiddev === false) throw new Exception("RAID device must be specified");
	}
} catch (Exception $e) {
	printf("ERROR: %s\n", $e->getMessage());
	echo "\n";
	echo "Usage:\n";
	echo "    expandable-raid.php --create [--level RAIDLEVEL] [--chunk CHUNKKB] [--layout LAYOUT] --vg VOLGROUP --partitions PART1,PART2,PART3\n";
	echo "    expandable-raid.php --extend --vg VOLGROUP --raid RAIDDEV --partition PART1\n";
	echo "    expandable-raid.php --remove --vg VOLGROUP --raid RAIDDEV\n";
	echo "\n";
	exit(1);
}



function run_command($cmd, $err_ok = false) {
	echo " + $cmd\n";
	
	if ($GLOBALS['dryrun']) return true;
	
	system($cmd, $retval);
	if ($retval > 0) {
		if ($err_ok) return false;
		throw new Exception("ERROR: command did not complete successfully!");
	}
	return true;
}


// Some common escaping...
$sh_volgroup = escapeshellarg($volgroup);


try {
	if ($mode === "create") {
		//************************** BEGIN CREATE MODE PREPARATION ****************************
		
		// Make sure any partitions being used in a volume group are moved off
		foreach ($partitions as $part) {
			$sh_part = escapeshellarg($part);
			if (run_command("pvdisplay $sh_part >/dev/null 2>/dev/null", true)) {
				echo "Removing $part from volume group...\n";
				run_command("pvmove --autobackup y $sh_part", true);
				run_command("vgreduce --autobackup y $sh_volgroup $sh_part");
				run_command("pvremove $sh_part");
			}
		}
		
		for ($i = 0; file_exists("/dev/md$i"); $i++);
		$raiddev = "/dev/md$i";
		
		//************************** END CREATE MODE PREPARATION ****************************
	} else if ($mode === "extend" || $mode === "remove") {
		//************************** BEGIN EXTEND MODE PREPARATION (OR REMOVE MODE) ****************************
		
		$old_partitions = false;
		$dev = substr($raiddev, 5);
		
		echo "Determining current partitions in $raiddev...\n";
		foreach (file("/proc/mdstat") as $line) {
			if (!preg_match("/^$dev : active raid([0-9]+) (.*)/", $line, $regs)) continue;
			
			$level = intval($regs[1]);
			if (!in_array($level, [ 0, 1, 5, 10 ])) {
				throw new Exception("ERROR: Unknown RAID level ($level)");
			}
			
			if (!preg_match_all('/([a-z]+[0-9]+)\[[0-9]+\]/', $regs[2], $matches)) {
				throw new Exception("ERROR: Could not get partition information from $regs[2]");
			}
			$old_partitions = $matches[1];
		}
		if ($old_partitions === false) {
			throw new Exception("ERROR: Could not get partition information for $raiddev");
		}
		
		// Get the chunk size for the existing RAID device
		$md = substr($raiddev, 5);
		if (file_exists("/sys/block/$md/md/chunk_size")) {
			$chunk = intval(file_get_contents("/sys/block/$md/md/chunk_size")) / 1024;
		}
		
		// Get the layout of the existing RAID device
		if (in_array($level, [ 5, 10 ])) {
			$output = shell_exec("mdadm --detail $raiddev");
			if (!preg_match('/Layout : (.*)$/m', $output, $regs)) throw new Exception("Could not determine layout of $raiddev");
			$layout = $regs[1];
			
			if ($level == 10) {
				list($layout_type, $layout_num) = explode("=", $layout);
				$layout_num = intval($layout_num);
				if ($layout_num < 1 || !in_array($layout_type, [ "near", "far", "offset" ])) {
					throw new Exception("$raiddev has an unknown layout: $layout");
				}
				$layout = $layout_type[0] . $layout_num;
			} else if ($level == 5) {
				if (!in_array($layout, [ "left-asymmetric", "left-symmetric", "right-asymmetric", "right-symmetric" ])) {
					throw new Exception("$raiddev has an unknown layout: $layout");
				}
			}
		}
		
		echo "Removing RAID device: $raiddev\n";
		run_command("pvmove --autobackup y $raiddev", true);
		run_command("vgreduce --autobackup y $sh_volgroup $raiddev");
		run_command("pvremove $raiddev");
		
		echo "Stopping RAID device $raiddev\n";
		run_command("mdadm --stop $raiddev");
		
		echo "Zeroing superblocks for old RAID drives\n";
		foreach ($old_partitions as $part) {
			$part = "/dev/$part";
			echo " + $part\n";
			run_command("mdadm --zero-superblock $part");
			$partitions[] = $part;
		}
		
		//************************** END EXTEND MODE PREPARATION ****************************
	}
	
	
	
	
	if ($mode === "create" || $mode === "extend") {
		//************************** BEGIN RAID DEVICE CREATION ****************************
		
		echo "Creating RAID device: $raiddev\n";
		
		$num_parts = count($partitions);
		$sh_partitions = "";
		foreach ($partitions as $part) {
			$sh_partitions .= escapeshellarg($part) . " ";
		}
		
		$other_options = "";
		if ($layout !== false) $other_options .= " --layout=".escapeshellarg($layout);
		
		run_command("mdadm --create --verbose $raiddev --level=$level --chunk=$chunk $other_options --raid-devices=$num_parts $sh_partitions");
		run_command("pvcreate $raiddev");
		run_command("vgextend $sh_volgroup $raiddev");
		
		//************************** END RAID DEVICE CREATION ****************************
	}
} catch (Exception $e) {
	printf("ERROR: %s\n", $e->getMessage());
	exit(2);
}

File: expandable-raid.pl
#!/usr/bin/env perl

use strict;
use warnings;
use Getopt::Long qw( VersionMessage );

#
# expandable-raid.pl
#
# Use Linux RAID and LVM together to add disks to your RAID array without
# stopping it.
#
# Github repositories
# https://github.com/stevecoug/expandable-raid  (PHP)
# https://github.com/daoswald/expandable-raid   (Perl rewrite)
#
# Copyright (c) 2013 Steve Meyers.
# Perl translation Copyright (c) 2013 Steve Meyers, David Oswald
#
# This Source Code Form is subject to the terms of the Mozilla Public License,
# v. 2.0. If a copy of the MPL was not distributed with this file, you can
# obtain one at http://mozilla.org/MPL/2.0/.
#

BEGIN { Getopt::Long::Configure('debug') if $ENV{EXPRAID_OPTIONS_DEBUG} }

our $VERSION = '1.0';

# CONVENTIONS: Constants are UPPER_CASED. Package Globals are $UPPER_CASED.
#              Lexicals are $lower_cased.

use constant {
    FALSE           => 0,
    TRUE            => 1,
    ERR_OK          => 1,
    DEFAULT_CHUNK   => 64,
    DEFAULT_LEVEL   => 5,
    VALID_LEVELS    => qr{^(?:0|1|5|10)$},
    VALID_RAID_DEV  => qr{^/dev/md[0-9]+$},
    MIN_VALID_CHUNK => 1,
    MAX_VALID_CHUNK => 1024,
};

our ( $MODE, $VOLGROUP, $RAIDDEV, $LAYOUT, $DRYRUN ) = (FALSE) x 5;
our ( $CHUNK, $LEVEL, @PARTITIONS ) = ( DEFAULT_CHUNK, DEFAULT_LEVEL, () );

$DRYRUN = TRUE if $ENV{EXPRAID_DEBUG};

GetOptions(
    'create|c'        => sub { $MODE     = 'create'; },
    'extend|x'        => sub { $MODE     = 'extend'; },
    'remove|rm'       => sub { $MODE     = 'remove'; },
    'volgroup|vg|v=s' => sub { $VOLGROUP = $_[1]; },
    'help|h|?' => sub { print usage(); exit(0); },
    'version' => sub { VersionMessage($VERSION); },
    'layout|l=s'             => \$LAYOUT,
    'dry-run|dryrun|debug|d' => \$DRYRUN,
    'partitions|p=s'         => sub {
        push @PARTITIONS, map {
            my $part = $_;
            $part =~ s{^([^/])}{/dev/$1};
            $part;
          } split /,/, $_[1];
    },
    'level|l=i' => sub {
        $LEVEL = $_[1];
        die "Invalid RAID level: $LEVEL\n" unless $LEVEL =~ VALID_LEVELS;
    },
    'raid|r=s' => sub {
        ( $RAIDDEV = $_[1] ) =~ s{^([^/])}{/dev/$1};
        die "Invalid RAID device: $RAIDDEV\n"
          unless $RAIDDEV =~ m{^/dev/md[0-9]+$};
    },
    'chunk=i' => sub {
        $CHUNK = $_[1];
        die "Chunk size must be between 1-1024.\n"
          if $CHUNK < MIN_VALID_CHUNK || $CHUNK > MAX_VALID_CHUNK;
    },
) or die usage();

warn "****    DRYRUN mode.    ****\n" if $DRYRUN;

eval {
    die "You must specify either --create or --extend\n" if !$MODE;
    die "Volume group must be specified\n"               if !$VOLGROUP;
    die "Partitions must be specified\n"
      if !@PARTITIONS && $MODE =~ m/^(?:create|extend)$/;
    die "RAID device must be specified\n" if $MODE eq 'extend' && !$RAIDDEV;
    1;    # Success.
} or die $@, usage();

my $sh_volgroup = php_escapeshellarg($VOLGROUP);

eval {
    for ($MODE) {
        /^(?:create)$       /x && do {
            $RAIDDEV = create_prep( \@PARTITIONS, $sh_volgroup );
        };
        /^(?:extend|remove)$/x && do {
            my @old_partitions = ();
            ( $LEVEL, $CHUNK, $LAYOUT, @old_partitions ) =
              extend_remove_prep( $RAIDDEV, $LEVEL, $CHUNK, $sh_volgroup );
            @PARTITIONS = (@old_partitions, @PARTITIONS);
        };
        /^(?:create|extend)$/x && do {
            create_raid( $RAIDDEV, $LAYOUT, $LEVEL, $CHUNK, $sh_volgroup,
                \@PARTITIONS );
        };
    }
    1;    # Success; no exceptions.
} or die "$@\n";


# ------------- actions ------------

sub create_prep {
    my ( $parts, $sh_vg ) = @_;
    foreach my $part ( @{$parts} ) {
        my $sh_part = php_escapeshellarg($part);
        if (
            run_command( "pvdisplay $sh_part >/dev/null 2>/dev/null", ERR_OK ) )
        {
            print "Removing $part from volume group...\n";
            run_command( "pvmove --autobackup y $sh_part", ERR_OK );
            run_command( "vgreduce --autobackup y $sh_vg $sh_part" );
            run_command( "pvremove $sh_part" );
        }
    }
    my $i = 0;
    $i++ while -e "/dev/md$i";
    return "/dev/md$i";
}


sub extend_remove_prep {
    my ( $raiddev, $level, $chunk, $sh_vg ) = @_;
    my @old_partitions;
    my $dev = substr( $raiddev, 5 );
    print "Determining current partitions in $raiddev...\n";
    
    open my $mdstat_fh, '<', '/proc/mdstat' or die $!;

    while ( my $line = <$mdstat_fh> ) {
        next unless $line =~ m/^$dev : active raid([0-9]+) (.*)/;
        my $partition_info;
        ( $level, $partition_info ) = ( $1, $2 );
        die "ERROR: Unknown RAID level ($level)\n"
          unless $level =~ m/^(?:0|1|5|10)$/;
        die "ERROR: Could not get partition information from $partition_info\n"
          unless @old_partitions
            = $partition_info =~ /([a-z]+[0-9]+)\[[0-9]+\]/g;
    }
    close $mdstat_fh;

    die "ERROR: Could not get partition information for $raiddev\n"
      unless @old_partitions;

    # Get the chunk size for the existing RAID device
    if ( -e "/sys/block/$dev/md/chunk_size" ) {
        open my $cs_fh, '<', "/sys/block/$dev/md/chunk_size" or die $!;
        $chunk = int( <$cs_fh> / 1024 );
        close $cs_fh;
    }

    # Get the layout of the existing RAID device
    my $layout;
    if ( $level =~ m/^(?:5|10)$/ ) {
        my $output = `mdadm --detail $raiddev`;
        die "ERROR: Could not determine layout of $raiddev\n"
          unless $output =~ m/Layout : (.*)$/m;
        $layout = $1;

        if ( $level == 10 ) {
            my ( $layout_type, $layout_num ) = split /=/, $layout;
            $layout_num = int $layout_num;
            die "ERROR: $raiddev has an unknown layout: $layout\n"
              if $layout_num < 1 || $layout_type !~ m/^(?:near|far|offset)$/;
            $layout = substr( $layout_type, 0, 1 ) . $layout_num;
        }
        elsif ( $level == 5 ) {
            die "ERROR: $raiddev has an unknown layout: $layout\n"
              unless $layout =~ m/^(?:left|right)-a?symmetric$/;
        }
    }
    print "Removing RAID device: $raiddev\n";
    run_command( "pvmove --autobackup y $raiddev", ERR_OK );
    run_command("vgreduce --autobackup y $sh_vg $raiddev");
    run_command("pvremove $raiddev");

    print "Stopping RAID device $raiddev\n";
    run_command("mdadm --stop $raiddev");

    print "Zeroing superblocks for old RAID drives\n";
    my @partitions;
    foreach my $part (@old_partitions) {
        $part = "/dev/$part";
        print " + $part\n";
        run_command("mdadm --zero-superblock $part");
        push @partitions, $part;
    }
    return ( $level, $chunk, $layout, @partitions );
}


sub create_raid {
    my ( $raiddev, $layout, $level, $chunk, $sh_vg, $parts_aref ) = @_;

    print "Creating RAID devicd $raiddev\n";

    my $num_parts     = @$parts_aref;
    my $sh_partitions = join q{ }, map { php_escapeshellarg($_) } @$parts_aref;
    my $other_options = q{};
    $other_options = " --layout=" . php_escapeshellarg($layout);

    run_command( "mdadm --create --verbose $raiddev --level=$level "
                 . "--chunk=$chunk $other_options --raid-devices=$num_parts "
                 . "$sh_partitions"
    );
    run_command("pvcreate $raiddev");
    run_command("vgextend $sh_vg $raiddev");
    return; # Nothing.
}

# ------------- utility functions ------------

sub usage {
    return << 'USAGE';

Usage:
    expandable-raid.pl --create [--level RAIDLEVEL] [--chunk CHUNKKB] [--layout LAYOUT] --vg VOLGROUP --partitions PART1,PART2,PART3 [--dryrun]
    expandable-raid.pl --extend --vg VOLGROUP --raid RAIDDEV --partition PART1 [--dryrun]
    expandable-raid.pl --remove --vg VOLGROUP --raid RAIDDEV [--dryrun]
    expandable-raid.pl --version
    expandable-raid.pl --help

Setting EXPRAID_DEBUG forces "--dryrun mode". 

USAGE
}

sub php_escapeshellarg {
    my $str = @_ ? shift : $_;
    $str =~ s/((?:^|[^\\])(?:\\\\)*)'/$1\\'/g;
    return "'$str'";
}

sub run_command {
    my ( $cmd, $err_ok ) = @_;
    print " + $cmd\n";
    return TRUE if $DRYRUN || system($cmd ) == 0;
    return FALSE if $err_ok;
    die "ERROR: Command did not complete successfully!\n";
}
